### DISCUSSION TITLE
Join/split pdf files?

### DISCUSSION THREAD

**Discussion by zubkovla** [sumatrapdfreader/sumatrapdf#3174] [Ideas]
Hi,
quite oftenly there is a need to join few pdf files into one. Presumably there are more efficient ways then just print append it via some pdf printer, time and space wise.
Sometimes there is a need to extract few pages.

Won't it be nice if Sumatra could deal with it so fluently and quickly as it views pdfs?

Thanks
/I've searched for keywords "split" and "join" and found nothing regarding this topic

**Comment by GitHubRulesOK** [comment]
reordering pages is a highly complex editor function since it is not just affecting viewing take just one simple page this is just its structure

```
1 0 obj<</Pages 2 0 R/Type/Catalog>>endobj

2 0 obj<</Count 1/Kids[3 0 R]/Type/Pages>>endobj

3 0 obj<</Contents 4 0 R/MediaBox [0 0 594 792]/Parent 2 0 R/Resources<</Font<</F1<</Type/Font/Subtype/Type1/BaseFont/Helvetica>>>>>>/Resources<</Font<</F2<</BaseFont /Symbol/Subtype/Type1/Type/Font>>>>>>/Type/Page>>endobj

4 0 obj<</Length 999>>
stream
```

looks simple just change the order of the children (here only a single obj 3) but note the fonts are part of this very simple description so adjusting page order needs adding and subtracting fonts and its a lot more complex than that as many other objects like annotation and images will also be referenced back to child 3 so the scale of changes with 2 pages is greater than 2 times the trouble.

PDF is a tangled web of inter related objects, hence editors are usually many times more code so they can rewrite the nest of resources.

**Comment by AnEnglishmanInNorway** [comment]
Split/Join (and other rearrangements) is certainly useful, but I'm not sure Sumatra is the place to do the job.
I use [pdfsam ](https://pdfsam.org/) for rearranging pdf pages. It's got more expensive since I started using it, but when I found it I was so happy with the result that I bought myself a perpetual license for the visual edition. Worth a look maybe.

**Reply by GitHubRulesOK** [reply]
pdfSAM is good but others are as good when slow see pdf arranger portable is full run time python + pike (qpdf in a wrapper) https://github.com/pdfarranger/pdfarranger

**Reply by zubkovla** [reply]
Actually I'd found and quite happy with online service "I love pdf"

**Reply by goyalyashpal** [reply]
> but others are as good when slow

can u rephrase?

**Reply by GitHubRulesOK** [reply]
@yashpalgoyal1304 
most slit and merge for free will be based on slower iterative run times Java or Python  or  other iterative language etc. Don't get me wrong they are powerful GUI solutions and fast when loaded but many dependencies, adds to run times.

**Comment by meteorquake** [comment]
I agree SumatraPDF isn't the place for such dedicated tasks. But there are free programs out there.
For just simple tasks like joining/splitting I use PdfTKBuilder - http://www.angusj.com/pdftkb/
For editing bookmarks and basic annotations I use Foxit Reader
For many other PDF tasks like OCR, watermarking, page marking etc I use PDF24 - https://www.pdf24.org

**Comment by goyalyashpal** [comment]
That was the very first instance of such cheap spam i have ever seen on github/gitlab etc. And that too, even in first go, two instances of it. 

Kinda strange that among all my subscibed repos, only sumatra's repo was targetted?

**Reply by GitHubRulesOK** [reply]
@yashpalgoyal1304 
SumatraPDF old forum was popular target for spam but not usually aimed at older generation ;-), much was filtered by me adding posting restrictions that worked some of the time, but here on GitHub we have to rely on flagging in appropriate content

**Reply by goyalyashpal** [reply]
lol i wonder why do ppl do such thing, what is their success rate (it would have been pretty small i'd guess) and also, how much do they gain on the small case of victims they manage to trap?

**Comment by Shohreh** [comment]
I was about to suggest this feature.

It's not uncommon for users to need to extract some pages of a longer PDF, and most don't have the skills/wish to look for another tool. So I agree it'd be a nice addition to SumatraPDF, especially since [MuPDF](https://mupdf.readthedocs.io/en/latest/C-API.html) — on which SumatraPDF apparently lies — supports this feature out of the box.

Joining looks harder to do, and is possibly a less common need.

**Comment by meteorquake** [comment]
I should add to my comment on PDFTK there's an enhanced version at https://pdftk-builder-enhanced.sourceforge.io/
So what I have on Sumatra under File menu is a VBS that pops up a file's location for copying (Ctrl-C)
```
		CommandLine = "C:\Windows\System32\wscript.exe" "C:\Scripts\right-clicks\locator.vbs" "%1"
		Name = Locator
		Filter = *.*
```
then I launch PDFTK (from SlickRun, but could also be added under File if wished) and adding the clipboard file tell it the pages I want, it would be ideal for PDFTK to have a command line switch that on opening adds a file, perhaps that's something to suggest to the author
It could also be that Sumatra could have a "copy filename of current file to clipboard" option (maybe it's present?)
I also have IrfanView added under File so I can click to open the current PDF on the current page in IrfanView where I can crop etc and save it as a PNG or JPG
```
		CommandLine = "C:\Program Files\IrfanView\i_view64.exe" "%1" /page=%p
		Name = IrfanView
		Filter = *.*
```
It may be Sumatra could automatically detect IrfanView and any other popular programs and offer to provide the menu link-in.
As mentioned I think these work well as separate dedicated programs outside Sumatra but Sumatra could streamline making them available, such as at installation or an option to detect and add them from the menu

**Comment by mikeybuhtig** [comment]
Another solution, free at the current time, is [PDFgear](https://www.pdfgear.com/). Besides split and join functions, it also can export/import to/from other formats, compress files, do OCR, fill forms, handle fill & sign, and password protect.

I like SumatraPDF as my quick viewer and PDFgear is the perfect adjunct for when I need to modify the files.

**Reply by GitHubRulesOK** [reply]
HMM will not run on my machine even with adding more than 350 files, but does show as has multiple google tracking files so I will give it a miss as free in this case comes at too heavy a price. There are much easier lighter alternatives for everyday edits of a pdf.

**Comment by Laidbak300** [comment]
I know this is a bit later but yes to @Shohreh. If MuPDF supports PDF extraction out of the box, it would be highly useful and there's little excuse not to implement it in Sumatra. Sumatra is aaaaaaaawesome but I need a whole 'nother PDF tool, PDFExchangeEdit, much bigger than Sumatra but a less capable reader, just for the extraction capability. That's all I use it for and I would love to get it off my machine--most of its capabilities are disabled behind the paywall, so it has a big footprint for this tiny single function it keeps free. 

But PDFExchangeEdit does happen to be really, really good at extraction, and it's 20 times more convenient than meteorquake's solution. Fabulous dialog with every extraction option you need, so I grimace at the size and install anyhow. When--and surely this has gotta be when, not if--you implement the MuPDF extraction capability, take a look at PDFExchangeEdit's extraction interface. You could do a lot worse than just straight-up copying it and all its options. At least make sure those options are available one way or another.

**Reply by GitHubRulesOK** [reply]
@Laidbak300
My PDF apps in order of use are SumatraPDF (Reading), and then equally Mutool for some Bulk Manipulation tasks and Tracker PDF Xchange Edit for precision (Writing). But none of those for re-ordering nor extraction.

PDF as a file format was never designed for editing in the way Word files can be done, and manipulation needs different profiling of tasks for each batch of changes.

Text extraction of glyphs as editable UTF text loses styling as styles are not present in a PDF printout. OCR is usually Glyphless Font
You may think it looks red or italic or bold, but they are in PDF simply morphing through "GraphicState"s or "Appearance" which is not really "Stylish".

PDF pages are stored in a semi-random state. Finding & Re-Ordering & Rendering a group from that jumble for display horizontally (or vertically), in reader screen sequence is the biggest slow down factors for PDF use. The more compressed pages it has to expand to screen, the slower it gets (don't use compression beyond the normal use for embedded images).

As I described above there is collateral damage when using split and merge, but re-ordering should not lose anything. The file size should be 100% identical. I have yet to see a PDF re-arranger that gets 1000 bytes and shuffles the 1000 bytes correctly, not 999 nor 1001.

**Reply by Laidbak300** [reply]
It may be true that PDFExchangeEdit doesn't get 100% accuracy with extraction, but as a user I can't tell. Its results work fine, so they're acceptable. If MuPDF can't replicate that result, however, then I guess I can understand why it's not in the cards for Sumatra.

**Reply by GitHubRulesOK** [reply]
@Laidbak300 
SumatraPDF is increasingly expanding on the internal MuPDF Tool abilities.
thus file content extraction functions more recently are enhanced
clipboard or text images and pages can be even edited in a variety of ways, not just all the historic ones.

This thread is about splitting and joining files 
- need to join few pdf files into one.
- is a need to extract few pages.

**Those are NOW covered by EXPORT as several parts or MERGE into one and those are current functions of pre-release so am closing the thread**

**Comment by goyalyashpal** [comment]
i didn't find any export / save-as / merge / join button-or-menu-or-option in the sumatrapdf-3.7 ?

**Reply by GitHubRulesOK** [reply]
@goyalyashpal you are correct there is no buttons that cover split and merge like dedicated editors do. which is why we get many I use this or that editor app (especially online links).

SumatraPDF has export pages in a user selected range 
<img width="624" height="188" alt="image" src="https://github.com/user-attachments/assets/c0cd49af-f902-4a63-bc88-4f957302e9ed" />
and also has command line merge OF those you chose https://www.sumatrapdfreader.org/docs/Tool-merge
